File: solutions/problem_0039_combination_sum/solution.py
"""
Problem 39: Combination Sum
===========================
Difficulty: Medium
Tags: Array, Backtracking
Pattern: Backtracking

Time Complexity:  O(n^(target/min))
Space Complexity: O(target/min)
"""

class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        result = []
        def backtrack(start, path, remaining):
            if remaining == 0:
                result.append(path[:])
                return
            for i in range(start, len(candidates)):
                if candidates[i] > remaining:
                    continue
                path.append(candidates[i])
                backtrack(i, path, remaining - candidates[i])
                path.pop()
        backtrack(0, [], target)
        return result

File: solutions/problem_0039_combination_sum/solution.cpp
/*
 * Problem 39: Combination Sum
 * ===========================
 * Difficulty: Medium
 * Tags: Array, Backtracking
 * Pattern: Backtracking
 *
 * Time Complexity:  O(n^(target/min))
 * Space Complexity: O(target/min)
 */

#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        // Backtracking - O(2^n) or O(n!) time
        vector<vector<int>> result;
        vector<int> path;
        function<void(int)> backtrack = [&](int start) {
            result.push_back(path);
            for (int i = start; i < (int)candidates.size(); i++) {
                path.push_back(candidates[i]);
                backtrack(i + 1);
                path.pop_back();
            }
        };
        backtrack(0);
        return result;
    }
};
